tempesta_fw/t/unit/sched_helper.c

```diff
@@ -133,3 +134,84 @@ test_conn_release_all(TfwSrvGroup *sg)
 		}
 	}
 }
+
+void test_sched_generic_empty_sg(struct TestSchedHelper * sched_helper)
+{
+	size_t i, j;
+	TfwSrvGroup *sg;
+
+	BUG_ON(!sched_helper);
+	BUG_ON(!sched_helper->scheduler);
+	BUG_ON(!sched_helper->connections_types);
+	BUG_ON(!sched_helper->get_scheduler_arg);
+	BUG_ON(!sched_helper->free_scheduler_arg);
+
+	sg = test_create_sg("test", sched_helper->scheduler);
+
+	for (i = 0; i < sched_helper->connections_types; ++i) {
+		for (j = 0; j < 3; ++j) {
+			TfwMsg * msg = sched_helper->get_scheduler_arg(i);
+			TfwConnection *conn = sg->sched->sched_srv(msg, sg);
+			EXPECT_NULL(conn);
+			sched_helper->free_scheduler_arg(msg);
+		}
+	}
+
+	test_sg_release_all();
+}
+
+void test_sched_generic_one_srv_in_sg_zero_conn(struct TestSchedHelper * sched_helper)
+{
+	size_t i, j;
+	TfwSrvGroup *sg;
+
+	BUG_ON(!sched_helper);
+	BUG_ON(!sched_helper->scheduler);
+	BUG_ON(!sched_helper->connections_types);
+	BUG_ON(!sched_helper->get_scheduler_arg);
+	BUG_ON(!sched_helper->free_scheduler_arg);
+
+	sg = test_create_sg("test", sched_helper->scheduler);
+
+	test_create_srv("127.0.0.1", sg);
+
+	for (i = 0; i < sched_helper->connections_types; ++i) {
+		for (j = 0; j < 3; ++j) {
+			TfwMsg * msg = sched_helper->get_scheduler_arg(i);
+			TfwConnection *conn = sg->sched->sched_srv(msg, sg);
+			EXPECT_NULL(conn);
+			sched_helper->free_scheduler_arg(msg);
+		}
+	}
+
+	test_sg_release_all();
+}
+
+void test_sched_generic_max_srv_in_sg_and_zero_conn(struct TestSchedHelper * sched_helper)
+{
+	size_t i, j;
+	TfwSrvGroup *sg;
+
+	BUG_ON(!sched_helper);
+	BUG_ON(!sched_helper->scheduler);
+	BUG_ON(!sched_helper->connections_types);
+	BUG_ON(!sched_helper->get_scheduler_arg);
+	BUG_ON(!sched_helper->free_scheduler_arg);
+
+	sg = test_create_sg("test", sched_helper->scheduler);
+
+	for (j = 0; j < TFW_SG_MAX_SRV; ++j) {
+		test_create_srv("127.0.0.1", sg);
+	}
+
+	for (i = 0; i < sched_helper->connections_types; ++i) {
+		for (j = 0; j < 2 * TFW_SG_MAX_SRV; ++j) {
```

Comment:
What is the point of trying twice as many times that there's servers in the group? Same question applies to the above two functions, where the test is tried three times.
There's similar code in other parts of this patch set.